controller/lib/hal/adc.cpp

```diff
@@ -0,0 +1,240 @@
+
+  // Perform a power-up and calibration sequence on
+  // the A/D converter
+  ADC_Regs *adc = ADC_BASE;
+
+  // Exit deep power down mode and turn on the
+  // internal voltage regulator.
+  adc->adc[0].ctrl = 0x10000000;
+
+  // Wait for the startup time specified in the datasheet
+  // for the voltage regulator to become ready.
+  // The time in the datasheet is 20 microseconds, but
+  // I'll wait for 30 just to be extra conservative
+  BusyWaitUsec(30);
+
+  // Calibrate the A/D for single ended channels
+  adc->adc[0].ctrl |= 0x80000000;
+
+  // Wait until the CAL bit is cleared meaning
+  // calibration is finished
+  while (adc->adc[0].ctrl & 0x80000000) {
+  }
+
+  // Clear all the status bits
+  adc->adc[0].stat = 0x3FF;
+
+  // Enable the A/D
+  adc->adc[0].ctrl |= 0x00000001;
+
+  // Wait for the ADRDY status bit to be set
+  while (!(adc->adc[0].stat & 0x00000001)) {
+  }
+
+  adc->adc[0].cfg1.dmaen = 1;
+  adc->adc[0].cfg1.dmacfg = 1;
+  adc->adc[0].cfg1.cont = 1;
+
+  // Enable ADC oversampling.
+  adc->adc[0].cfg2.rovse = (oversample_log2 > 0) ? 1 : 0;
+
+  // The over sample setting is the log base 2 of the number of samples
+  // less one.
+  // For example, for 16 samples we would use log2(16)-1 = 3
+  adc->adc[0].cfg2.ovsr = (oversample_log2 > 0) ? oversample_log2 - 1 : 0;
+
+  // The A/D data register can only return 16 bits, so we could get
+  // truncated data if our oversampling is higher then 16.  In this
+  // case I need to tell the hardware to downshift the sum
+  adc->adc[0].cfg2.ovss = (oversample_log2 < 4) ? 0 : (oversample_log2 - 4);
+
+  // Set sample time. I'm using 92.5 A/D clocks to sample.
+  static_assert(adc_samp_time == 92);
+  adc->adc[0].samp.smp5 = 5;
+  adc->adc[0].samp.smp6 = 5;
+  adc->adc[0].samp.smp9 = 5;
+  adc->adc[0].samp.smp15 = 5;
+
+  // Set the conversion sequence
+  adc->adc[0].seq.len = adc_channels - 1;
+  adc->adc[0].seq.sq1 = 6;
+  adc->adc[0].seq.sq2 = 9;
+  adc->adc[0].seq.sq3 = 15;
+
+  // I use DMA1 channel 1 to copy my A/D readings into my buffer
+  EnableClock(DMA1_BASE);
+  DMA_Regs *dma = DMA1_BASE;
+  int C1 = static_cast<int>(DMA_Chan::C1);
+
+  dma->channel[C1].pAddr = &adc->adc[0].data;
+  dma->channel[C1].mAddr = adc_buff;
+  dma->channel[C1].count = adc_samp_history * adc_channels;
+
+  dma->channel[C1].config.enable = 0;
+  dma->channel[C1].config.tcie = 0;
+  dma->channel[C1].config.htie = 0;
+  dma->channel[C1].config.teie = 0;
+  dma->channel[C1].config.dir =
+      static_cast<REG>(DmaChannelDir::PERIPHERAL_TO_MEM);
+  dma->channel[C1].config.circular = 1;
+  dma->channel[C1].config.perInc = 0;
+  dma->channel[C1].config.memInc = 1;
+  dma->channel[C1].config.psize = 1;
+  dma->channel[C1].config.msize = 1;
+  dma->channel[C1].config.priority = 0;
+  dma->channel[C1].config.enable = 1;
+
+  // Start the A/D converter
+  adc->adc[0].ctrl |= 4;
+}
+
+// Read the specified analog input.
+Voltage HalApi::analogRead(AnalogPin pin) {
```

Comment:
Would it be even better to read this array once, getting values for all three entries in parallel?  If so, could we add a TODO to change the HAL API to allow that?

Reply:
I don't think it actually makes a significant difference from a timing standpoint.  It would be a touch faster, but the array we iterate through should be small.  I think the current implementation should be faster then the previous one because we aren't waiting for the A/D to do 32 conversions when we call this.